Summary of changes

c0nwild opened the pull request:
This PR mainly enables the solver to use the Equation package. Part of the baselining effort.

calculator.py: Data container as only parameter for class constructor, calculate function now only resolves autocalc methods and calls Expression. Makes use of Formula class.
data_container.py: Data container is build by XMLReader class and contains all necessary information for performing the calculation
definitions.py: Contains generic type definitions
formula.py: Makes use of the expression class and creates the executable mathematical expression.
xmlreader.py: Does most of the work by analyzing the xml file created by VP and builds the data container.
main.py: binds together all of the subclasses. Is responsible for the recursive call of the calculation procedure to realize calculation of nested constraint properties.

Acceptance criteria:
- Error free execution
- PEP8 styleguide followed.

c0nwild committed:
Summary of changes

calculator.py: Data container as only parameter for class constructor, calculate function now only resolves autocalc methods and calls Expression. Makes use of Formula class.
data_container.py: Data container is build by XMLReader class and contains all necessary information for performing the calculation
definitions.py: Contains generic type definitions
formula.py: Makes use of the expression class and creates the executable mathematical expression.
xmlreader.py: Does most of the work by analyzing the xml file created by VP and builds the data container.
main.py: binds together all of the subclasses. Is responsible for the recursive call of the calculation procedure to realize calculation of nested constraint properties.

c0nwild committed:
Added several classes to xml_analyzer namespace.

c0nwild committed:
Corrections in import statements.

elheck commented:
Please use pylint with a pylintrc file and make the suggested changes

elheck reviewed (changes requested):
Please use pylint with a pylintrc file and make the suggested changes

elheck commented on main.py:
```diff
@@ -15,19 +14,18 @@
 PROJECT_FILE = ''
```
the project file is given per commandline why still have this?

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/4c16266a6748f824b112cf3f8e9ee007f8f268ff

elheck commented on main.py:
```diff
@@ -15,19 +14,18 @@
-def do_calculation(constraint_property_id: str, xmlreader: XMLReader) -> str:
-    constraint_property = xmlreader.find_constraint_property(constraint_property_id)
-
+def do_calculation(constraint_property_id: str, xmlreader: XMLReader) -> [str, float]:
```
do_claculation is the equivalent of "tu rechnen". Why not name it calculate or similiar

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/4c16266a6748f824b112cf3f8e9ee007f8f268ff

elheck commented on main.py:
```diff
@@ -15,19 +14,18 @@
-def do_calculation(constraint_property_id: str, xmlreader: XMLReader) -> str:
-    constraint_property = xmlreader.find_constraint_property(constraint_property_id)
-
+def do_calculation(constraint_property_id: str, xmlreader: XMLReader) -> [str, float]:
```
How can this be a str

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/4c16266a6748f824b112cf3f8e9ee007f8f268ff

elheck commented on xml_analyzer/__init__.py:
```diff
@@ -1,6 +1,8 @@
+from .calculator import Calculator
 
-__all__ = ['Calculator', 'XMLReader', 'Formula']
+__all__ = ['Calculator', 'XMLReader', 'Formula', 'AutocalcMethods', 'EquationTypes', 'DataContainer']
```
why is every module from this package publicly exposed. Only Calculator and XMLReader is used in main

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/f1272272794129d7e2ff4e59f447510111bf27f4

elheck commented on xml_analyzer/calculator.py:
```diff
@@ -1,63 +1,25 @@
-
-        if autocalc_method == 'automult':
-            pass
+            return round(sum(values), 3)
```
does only implicitly return without a return type if autocalc_method != "autosum"

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/d81255db0ad47a8137f453293869dd104d9eaaf7

elheck commented on xml_analyzer/calculator.py:
```diff
@@ -1,63 +1,25 @@
-        self.constraint_property = constraint_property
-        self.property_value_mapping = property_value_mapping
+    def __init__(self, data: DataContainer):
+        self.data = data
```
why is data public

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/d69fb3af4e8f294edfd506a5df7eb7ff6b81aca4

elheck commented on xml_analyzer/calculator.py:
```diff
@@ -1,63 +1,25 @@
-        self.property_value_mapping = property_value_mapping
+    def __init__(self, data: DataContainer):
+        self.data = data
+        self.formula = Formula(data.constraint_specification, data.get_variable_names())
```
why is formula public

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/d69fb3af4e8f294edfd506a5df7eb7ff6b81aca4

elheck commented on xml_analyzer/calculator.py:
```diff
@@ -1,63 +1,25 @@
-
-    def resolve_auto_calc_function(self, autocalc_method, type_name):
-        retval = None
+    def _resolve_auto_calc_functions(autocalc_method, values):
```
Type hint missing

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/d69fb3af4e8f294edfd506a5df7eb7ff6b81aca4

elheck commented on xml_analyzer/calculator.py:
```diff
@@ -1,63 +1,25 @@
 
-        return retval
+    def calculate(self) -> float:
+        for autocalc_method in self.data.mappings['autocalc']:
```
handle string keys like 'autocalc' like a magic number. If the name changes you have to change it everywhere you used it

c0nwild replied:
Fixed

c0nwild replied:
Better solution with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/56aed8eba9b4acceccf71949cee0658b3197752c

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -6,21 +6,31 @@ def __init__(self):
         self.mapping_element = collections.namedtuple('property_value_mapping', ['property', 'value'])
```
declaration of a type is not very nice in __init__. Rather handle it just like a class.
MappingElement = collections.namedtuple('property_value_mapping', ['property', 'value'])

outside of the class and then you can use it as following.


map_element = MappingElement(1,2)

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/6614ef2d31e5e3284ee434bbd113e446107a9606

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -6,21 +6,31 @@ def __init__(self):
         self.mapping_element = collections.namedtuple('property_value_mapping', ['property', 'value'])
         self.mappings = dict()
```
Are all of these instance attributes really public, if yes then why?

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/6614ef2d31e5e3284ee434bbd113e446107a9606

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -6,21 +6,31 @@ def __init__(self):
-        self.mappings['variables'][prop] = value
+        self.mappings['variables'][prop] = float(value)
 
     def add_result_property(self, prop):
```
type hint

elheck replied:
Also have a look into @ property this is what you want here

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/6614ef2d31e5e3284ee434bbd113e446107a9606. Great hint!!

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -6,21 +6,31 @@ def __init__(self):
+    def get_result_property(self) -> str:
+        return self.result_property
+
+    def add_dependency_mapping(self, prop: str):
```
make it a property as well

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/6614ef2d31e5e3284ee434bbd113e446107a9606

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -6,21 +6,31 @@ def __init__(self):
+            map_elements = self.mappings['autocalc'][autocalc_method]
+        except KeyError:
+            map_elements = list()
         map_element = self.mapping_element(sysml_type_name, values)
```
maybe put the rest of the function into an else block. Exception handling in python consists of try - except - else -finally (else and finally are optional)

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/6614ef2d31e5e3284ee434bbd113e446107a9606

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -6,21 +6,31 @@ def __init__(self):
+        map_elements.append(map_element)
+        self.mappings['autocalc'][autocalc_method] = map_elements
 
     def add_dependencies(self, dependencies: list):
```
property as well

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/6614ef2d31e5e3284ee434bbd113e446107a9606

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -29,7 +39,7 @@ def set_constraint_specification(self, constraint_spec: str):
         self.constraint_specification = constraint_spec
 
     def get_variable_names(self) -> list:
```
property as well - all setters and getters should be properties

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/6614ef2d31e5e3284ee434bbd113e446107a9606

elheck commented on xml_analyzer/formula.py:
```diff
@@ -1,13 +1,38 @@
-    def make_expression(self, expression: str):
-        return Expression(self.expression, self.variables)
+    def _make_expression(self, expression: str, variables: list):
+        expr_type, expr_sep = self._get_expression_type(expression)
```
It doesn't hurt to name it expression_type, expression_separator .

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/94c7a5274c3d122e82bf28a3421db92e9d3cfdd1

elheck commented on xml_analyzer/formula.py:
```diff
@@ -1,13 +1,38 @@
-        return Expression(self.expression, self.variables)
+    def _make_expression(self, expression: str, variables: list):
+        expr_type, expr_sep = self._get_expression_type(expression)
+        if expr_type == EquationTypes.equation:
```
From the naming alone I would think that you are comparing expression types to equation types and this seems fishy

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/94c7a5274c3d122e82bf28a3421db92e9d3cfdd1

elheck commented on xml_analyzer/formula.py:
```diff
@@ -1,13 +1,38 @@
+        return Expression(expression, variables)
+
+    @staticmethod
+    def _get_expression_type(expr: str) -> [EquationTypes, str]:
```
expression is not much longer than expr. Especially not longer than ex STRG+LEERTASTE ENTER

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/94c7a5274c3d122e82bf28a3421db92e9d3cfdd1

elheck commented on xml_analyzer/formula.py:
```diff
@@ -1,13 +1,38 @@
+        if expr_type == EquationTypes.equation:
+            expression = expression.split(expr_sep)[1]
 
+        return Expression(expression, variables)
```
Is it safe to make a Expression object with "undef" as first argument?

c0nwild replied:
Fixed.

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -1,38 +1,37 @@
 
     def __init__(self, project_file):
         self._project_file = project_file
         self.tree = Et.parse(self._project_file)
```
why is the project file private, but not the tree

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/125a9de3f5317b3cc4f907790a9fabccc9134c0e

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -71,21 +70,31 @@ def resolve_dependencies(self, binding_connector_list: list) -> list:
+
+        return element
+
+    def build_data_container(self, constraint_property_id: str) -> DataContainer:
```
method is very long you can easily factor out the for loop

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/125a9de3f5317b3cc4f907790a9fabccc9134c0e

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -1,38 +1,37 @@
 
-    def resolve_bindings(self, constraint_property: Et.Element) -> list:
+    def _resolve_bindings(self, constraint_property: Et.Element) -> list:
         query = ".//SysMLConstraintProperty[@Id='{}']//SysMLConstraintBlock".format(constraint_property.get("Id"))
```
Hm . In a perfect world all of the fragments of these xpath strings would be refectored into variables and a builder. I can accept the strings here

c0nwild replied:
This will be part of the next PR.

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -1,38 +1,37 @@
+                autocalc_sysml_type_name.append(method.name)
+                for occurrence in constraint_spec.split(method.name)[1:]:
                     search_param = occurrence[occurrence.find('(') + 1:occurrence.find(')')]
                     if search_param != '':
```
Can be simplified into 
`if search_param:
`

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/125a9de3f5317b3cc4f907790a9fabccc9134c0e

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -1,38 +1,37 @@
         return elements
 
-    def resolve_bindings(self, constraint_property: Et.Element) -> list:
+    def _resolve_bindings(self, constraint_property: Et.Element) -> list:
```
Instead of using list as type hint, you can use the metatype that was built for type hints.

`from typing import List`

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/125a9de3f5317b3cc4f907790a9fabccc9134c0e

c0nwild committed:
First bunch of changes to satisfy pylinter. Interim evaluation 9.82

c0nwild committed:
main.py:
- Changed function name
- Corrected type hint of function

c0nwild committed:
Changes to __init__.py:
- Only Calculator and XMLReader class will be publicly exposed.

c0nwild committed:
data_container.py:
- mapping dict fields are now defined in enum
- all attributes are now private
- getters and setters declared with property decorator
- added 'finally' statement in try-except statement

calculator.py:
- replaced hardcoded dict key

definitions.py:
- MappingElement defined here, in case it is used in more than one file

xmlreader.py, main.py:
- adaptions due to the changes described above

c0nwild committed:
Changes to xmlreader.py

- made class attributes private
- changed list typehint to metatype
- some code simplifications
- factored out code chunks from function build_data_container

c0nwild committed:
Changes due to linter findings

- refactored formula class to inherit from Equation class
- corrected order of import statements
- line too long

c0nwild committed:
Missed change requests

- made attributes of Calculator class private
- Added typehint

c0nwild committed:
Changes to formula.py

- renamed variables to make more sense
- renamed methodname to make more sense

c0nwild committed:
Changes to DataContainer

- refactored properties to return specific part of the mapping structure
- changes in calculator.py according to new properties

elheck commented on xml_analyzer/calculator.py:
```diff
@@ -1,63 +1,26 @@
+        for autocalc_method in self._data.autocalc_mappings:
+            for autocalc_entry in self._data.autocalc_mappings[autocalc_method]:
+                sysml_type_name = autocalc_entry.property
+                auto_value = self._resolve_auto_calc_functions(autocalc_method, autocalc_entry.value)
```
Return value None is not tested

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/7ff4099ad757bf5d465260c8ac8cb42df12253f0

elheck commented on main.py:
```diff
@@ -42,7 +37,7 @@ def main():
     # Get all constraint property ids in a package as list
     cp_ids_list = xmlreader.find_constraint_property_ids()
```
something_something_list is like hungarian notation. It is cleaner to just use the plural to indicate an iterable. So I am proposing: cp_ids_list -> cp_ids

c0nwild replied:
Whoops, overlooked it... Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/f9dc97172f7daab018167f85fdc90c1bf934e9f2

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -1,38 +1,72 @@
+    autocalc = 'autocalc'
 
 
 class DataContainer:
```
could be a [dataclasss](https://docs.python.org/3/library/dataclasses.html)

c0nwild replied:
Interesting. Changed to data class, but what is the advandage over enum class in this case?

c0nwild replied:
Changed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/a01925bac39c4c81ffc60b598221bd35835a36d1

elheck replied:
I was talking about DataContainer class

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/781e990bc4772f55cfbfba95834ee2ba5961f1b5

elheck commented on xml_analyzer/data_container.py:
```diff
@@ -1,38 +1,72 @@
+            map_elements = self._mappings[DataContainerFields.autocalc][autocalc_method]
+        except KeyError:
+            map_elements = list()
+        finally:
```
Everything in the finally block gets executed, even if a exception is raised. Is this intended? If you want to do something if no exception was raised, then use the else block

c0nwild replied:
I am using the KeyError exception to trigger the creation of an empty list in case there is not already an existing one for this specific key.... maybe this is a little too brutal?

elheck replied:
Thats fine, it just looked unintended

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -1,144 +1,153 @@
 class XMLReader:
-    autocalc_methods = ['autosum']
 
     def __init__(self, project_file):
```
A type hint for project file would be nice. You can indicate if it just the path (a str) or an open file - file object

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/729e6c63dfd4ab70a84ea5ce1e2c3dbe515eb5f1

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -1,144 +1,153 @@
-        for binding_connector in binding_connector_list:
+        for binding_connector in binding_connectors:
             stereo = binding_connector.find('./Stereotypes/Stereotype[@Name="external"]')
             if stereo is not None:
```
can be simplified to if stereo

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/ffeb519bb77b6dd72bd2e24167623a3a70184ea1

c0nwild replied:
Turns out this does not work. Reverted to former state with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/5e9632342512a874b9746546b01a3087305c6dd9

elheck commented on xml_analyzer/xmlreader.py:
```diff
@@ -1,144 +1,153 @@
+            val = self._root.find(".//*[@Id='{}']".format(id_from)).get('InitialValue')
+            prop = self._root.find(".//*[@Id='{}']".format(id_to)).get('Name')
             stereo = binding_connector.find('./Stereotypes/Stereotype[@Name="external"]')
             if stereo is not None:
```
could be simplified

c0nwild replied:
Fixed with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/ffeb519bb77b6dd72bd2e24167623a3a70184ea1

c0nwild replied:
Turns out this does not work. Reverted to former state with https://github.com/Squiddrone/VPParametricsSolver/pull/2/commits/5e9632342512a874b9746546b01a3087305c6dd9

c0nwild committed:
Changes to main.py

- Refactored variable name

c0nwild committed:
Changes to calculator.py

- auto_value variable checked for value

c0nwild committed:
Changes to data_container.py

- Changed enum class to data class

c0nwild committed:
Changes to xmlreader.py

- added type hint to init function

c0nwild committed:
Changes to xmlreader.py

- simplified if statement

c0nwild committed:
Changes to xmlreader.py

- simplified if statement does not work... reverted to former state

c0nwild committed:
Changes to data_container.py

- removed dataclass decorator
- changed class back to enum class
